compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUOps.cpp

```diff
@@ -194,29 +194,57 @@ MutableOperandRange CoalescedGatherDMAOp::getDpsInitsMutable() {
 
 // ParallelCombiningOpInterface implementation
 MutableOperandRange CoalescedGatherDMAOp::getUpdatedDestinations() {
+  // Only relevant for tensor operands
+  if (!isa<RankedTensorType>(getInit().getType())) {
+    return MutableOperandRange(getOperation(), /*start=*/0, /*length=*/0);
+  }
   // Return the init operand as the destination being updated
   return getInitMutable();
 }
 
 Operation *CoalescedGatherDMAOp::getIteratingParent() {
+  // Only relevant for tensor operands
+  if (!isa<RankedTensorType>(getInit().getType())) {
+    return nullptr;
+  }
   // Return the parent scf.forall operation
   return getOperation()->getParentOfType<scf::ForallOp>();
 }
 
 LogicalResult CoalescedGatherDMAOp::verify() {
-  auto initType = cast<RankedTensorType>(getInit().getType());
-  auto resultType = cast<RankedTensorType>(getResult().getType());
+  auto initType = getInit().getType();
+  auto resultType = getResult().getType();
 
   // Verify that this op is nested within an InParallelOpInterface op
-  if (!isa_and_nonnull<InParallelOpInterface>(getOperation()->getParentOp())) {
-    return emitOpError("must be nested within an operation implementing "
-                       "InParallelOpInterface");
+  // Note: This constraint only applies when working with tensors
+  if (isa<RankedTensorType>(initType)) {
+    if (!isa_and_nonnull<InParallelOpInterface>(
+            getOperation()->getParentOp())) {
+      return emitOpError("must be nested within an operation implementing "
+                         "InParallelOpInterface when using tensor operands");
+    }
   }
 
   if (initType != resultType) {
     return emitOpError("init and result must have the same type and shape");
   }
 
+  // Ensure all operands are either all tensors or all memrefs
+  bool hasTensor = isa<RankedTensorType>(initType);
+  bool hasMemRef = isa<MemRefType>(initType);
+
+  if (hasTensor) {
+    if (!isa<RankedTensorType>(getIndices().getType()) ||
+        !isa<RankedTensorType>(getSource().getType())) {
+      return emitOpError("all operands must be tensors when init is a tensor");
+    }
+  } else if (hasMemRef) {
+    if (!isa<MemRefType>(getIndices().getType()) ||
```

Comment:
The end goal is to vectorize the indices. If the plan is to bufferize -> unroll -> mem2reg away the private alloc for the indices producer that's ok, but I would have thought the easier path is to allow taking vector operands for the indices operand.

Reply:
The plan is indeed `bufferize -> unroll -> load_from_mem_to_vector`.

I was thinking expanding `vector.load` of indices can reuse the loop indices computation.